### PR TITLE
Problem: no test framework, no unit tests

### PR DESCRIPTION
**Pull request by sigiesec**
Solution: integrate unity, create basic unit test structure

- [x] adapt autotools build (@bluca?)
- [x] decide on directory structure: should API tests be parallel to unittests?

### CODE REVIEW

File: external/unity/version.txt [hunk 1]
```diff
@@ -0,0 +1 @@
+https://github.com/ThrowTheSwitch/Unity/commit/b4aca70fd9e0ddf0afbdafb1b826f5edcfc1049b
```
Comment (bluca) [inline external/unity/version.txt]: wouldn't it be better to sync from a release tag?

**Comment by bluca** [comment]
I can take care of autotools, but over the weekend (I hope :-) )

File: tests/unittests/unittest_poller.cpp [hunk 2]
```diff
@@ -0,0 +1,47 @@
+/*
+Copyright (c) 2007-2016 Contributors as noted in the AUTHORS file
```
Comment (bluca) [inline tests/unittests/unittest_poller.cpp]: date range should be just 2018 - sorry but with these things it's better to be pedantic...

**Comment by bluca** [comment]
I'm fine with having a unittest subdirectory of tests - it's better to keep them together

**Comment by sigiesec** [comment]
I asked myself if the directory structure should be changed to have tests/unittests and tests/apitests (or some other name, containing the existing tests). The current structure is problematic with cmake: properties are inherited from the base directory.

**Comment by bluca** [comment]
The problem is that moving all the files has the usual problems with git blame, and also there might be scripts and such that are broken and will annoy people, I'd rather leave them there if possible

**Comment by bluca** [comment]
I've added autotools support for unity and the unittests.
For the latter, for now, it's linking with libzmq.a - the alternative would be to list individually all the required source files in each Makefile.am target.

**Comment by sigiesec** [comment]
Cool, great :) Yes, that's what I am doing with cmake as well (link with libzmq-static). It would be quite hard to figure out which cpp files are really needed for each test, and I don't think that would provide a substantial benefit.

**Comment by bluca** [comment]
Ok - still needs some work though, check works but distcheck doesn't

**Comment by bluca** [comment]
distcheck should work fine now - also I rebased for the other changes to autotools, and added the copyright notice for debian packaging

**Comment by bluca** [comment]
One of the problems is that the static library build might be disabled, and then I think the build will fail. Should probably check for it.

**Comment by sigiesec** [comment]
Hm, yes... so maybe skip building the unit tests if the static library building is diabled? Maybe output a warning in that case. I think that is acceptable.

**Comment by sigiesec** [comment]
Regarding the directory structure: I see the problem with moving the existing files. What about placing unittests not inside tests but parallel to tests? cmake doesn't really go well with nesting directories, which do not inherit all properties. I need to fix this within unittests/CMakeLists.txt at the moment.

**Comment by bluca** [comment]
Fine for both - if you do that change, then I'll take care of the static thing in autotools

**Comment by bluca** [comment]
@sigiesec  one of the windows tests fail:

```
c:\projects\libzmq\tests\testutil_security.hpp(558): warning C4550: expression evaluates to a function which is missing an argument list (C:\projects\libzmq\tests\test_security_curve.cpp) [C:\projects\build_libzmq\tests\test_security_curve.vcxproj]
C:\projects\libzmq\tests\test_security_curve.cpp(53): error C3861: 'snprintf': identifier not found [C:\projects\build_libzmq\tests\test_security_curve.vcxproj]
```

Could you please have a look?

**Comment by sigiesec** [comment]
@bluca fixed the build failure with VS2013, hopefully

**Comment by sigiesec** [comment]
@bluca ok, I changed the directory structure as proposed, and let the unit tests build only if BUILD_STATIC is enabled (in cmake)

**Comment by bluca** [comment]
Great, done the same with autotools, also now ships the unittests/CMakeList.txt in the distributable tarball.

Also fixed the copyright dates in the new unit tests files

**Comment by bluca** [comment]
Another TODO (but can be done separately) I think would be to split the security tests, as they are a bit hybrid right now and include the source files - move those to the unittests, and leave API-only stuff in tests/

**Comment by bluca** [comment]
Also I've tested this branch with Unity on mips/armhf/arm64/hurd and it's working fine, later next week I'll do a few more architectures and Solaris 10

File: tests/unittests/unittest_poller.cpp (hunk 2 above)
Comment (sigiesec) [reply tests/unittests/unittest_poller.cpp]: Ok, thanks for updating this!

File: external/unity/version.txt (hunk 1 above)
Comment (sigiesec) [reply external/unity/version.txt]: Generally yes, I vaguely remember there was some reason not to use the latest release, but I will check again.

**Comment by bluca** [comment]
So what's left to do for this to be merged?

**Comment by sigiesec** [comment]
I think it can be merged :)

**Comment by bluca** [comment]
Yay, great work!

What about the clang-format test? Should we apply that diff?

**Pull request merged**